packages/agent-builder/src/defaults.ts

```diff
@@ -1543,15 +1627,93 @@ export const mastra = new Mastra({
+
+    // If no files specified, use legacy CLI-based validation for backward compatibility
+    if (!files || files.length === 0) {
+      return this.validateCodeCLI({ projectPath, validationType });
+    }
+
+    // Hybrid validation approach for specific files (default behavior)
+    for (const filePath of files) {
+      const absolutePath = isAbsolute(filePath) ? filePath : resolve(targetProjectPath, filePath);
+
+      try {
+        const fileContent = await readFile(absolutePath, 'utf-8');
+        const fileResults = await this.validateSingleFileHybrid(
+          absolutePath,
+          fileContent,
+          targetProjectPath,
+          validationType,
+        );
+
+        errors.push(...fileResults.errors);
+
+        // Track validation results
+        for (const type of validationType) {
+          const hasErrors = fileResults.errors.some(e => e.type === type && e.severity === 'error');
+          if (hasErrors) {
+            if (!validationsFailed.includes(type)) validationsFailed.push(type);
+          } else {
+            if (!validationsPassed.includes(type)) validationsPassed.push(type);
+          }
+        }
+      } catch (error) {
+        errors.push({
+          type: 'typescript',
+          severity: 'error',
+          message: `Failed to read file ${filePath}: ${error instanceof Error ? error.message : String(error)}`,
+          file: filePath,
+        });
+        validationsFailed.push('types');
+      }
+    }
+
+    const totalErrors = errors.filter(e => e.severity === 'error').length;
+    const totalWarnings = errors.filter(e => e.severity === 'warning').length;
+    const isValid = totalErrors === 0;
+
+    return {
+      valid: isValid,
+      errors,
+      summary: {
+        totalErrors,
+        totalWarnings,
+        validationsPassed,
+        validationsFailed,
+      },
+    };
+  }
+
+  /**
+   * CLI-based validation for when no specific files are provided
+   */
+  static async validateCodeCLI({
+    projectPath,
+    validationType,
+  }: {
+    projectPath?: string;
+    validationType: Array<'types' | 'lint' | 'schemas' | 'tests' | 'build'>;
+  }) {
+    const errors: Array<{
+      type: 'typescript' | 'eslint' | 'schema' | 'test' | 'build';
+      severity: 'error' | 'warning' | 'info';
+      message: string;
+      file?: string;
+      line?: number;
+      column?: number;
+      code?: string;
+    }> = [];
+    const validationsPassed: string[] = [];
+    const validationsFailed: string[] = [];
+
     const execOptions = { cwd: projectPath };
 
-    // TypeScript validation
+    // TypeScript validation (legacy approach)
     if (validationType.includes('types')) {
       try {
-        const fileArgs = files?.length ? files : [];
-        // Use execFile for safe argument passing to avoid shell interpretation
-        const args = ['tsc', '--noEmit', ...fileArgs];
-        await execFile('npx', args, execOptions);
+        const tscCommand = 'npx tsc --noEmit';
```

Comment:
I think a lot of these changes were to avoid shell interpretation. DId we want to change this back?